### ISSUE TITLE
Uploading large files to AWS with aws_s3_bucket_object

### ISSUE THREAD

**Issue by NathanDotTo** [hashicorp/terraform-provider-aws#19336]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
terraform -v
2021-05-12T09:50:18.204+0200 [DEBUG] Adding temp file log sink: /var/folders/18/h1984hd95ms_l22lvpwzc57r0000gn/T/terraform-log131399222
2021-05-12T09:50:18.204+0200 [INFO]  Terraform version: 0.15.3
2021-05-12T09:50:18.204+0200 [INFO]  Go runtime version: go1.16.3
2021-05-12T09:50:18.204+0200 [INFO]  CLI args: []string{"/usr/local/homebrew/bin/terraform", "-v"}
2021-05-12T09:50:18.205+0200 [TRACE] Stdout is a terminal of width 174
2021-05-12T09:50:18.205+0200 [TRACE] Stderr is a terminal of width 174
2021-05-12T09:50:18.205+0200 [TRACE] Stdin is a terminal
2021-05-12T09:50:18.205+0200 [DEBUG] Attempting to open CLI config file: /Users/nsowatsk/.terraformrc
2021-05-12T09:50:18.205+0200 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2021-05-12T09:50:18.207+0200 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2021-05-12T09:50:18.207+0200 [DEBUG] ignoring non-existing provider search directory /Users/nsowatsk/.terraform.d/plugins
2021-05-12T09:50:18.207+0200 [DEBUG] ignoring non-existing provider search directory /Users/nsowatsk/Library/Application Support/io.terraform/plugins
2021-05-12T09:50:18.207+0200 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2021-05-12T09:50:18.208+0200 [INFO]  CLI command args: []string{"version", "-v"}
Terraform v0.15.3
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.39.0
```

### Affected Resource(s)

aws_s3_bucket_object

### Terraform Configuration Files

```hcl
resource "aws_s3_bucket_object" "object" {
  bucket = "dnac-ova"
  key    = "dnac/dnac.ova"
  source = "/Users/x/Virtual Machines.localized/dnac.ova"

  # See Re: [hashicorp/terraform] Uploading a large file to AWS with Resource: aws_s3_bucket_object (#28678)
  # etag = filemd5("/Users/x/Virtual Machines.localized/dnac.ova")
}
```

Note that dnac.ova is ~80GB

Also note that this command works, in that the `test.txt` file appears in the bucket.

```
echo test > test.txt; aws s3api put-object --bucket dnac-ova --key test/test.txt --region us-east-1 --body test.txt
```

As a further test, I also tried:

```
resource "aws_s3_bucket_object" "object" {
  bucket = "dnac-ova"
  key    = "test2/test2"
  source = "test2.txt"
}
```

That also worked as expected, in that the file was uploaded to the bucket.

### Debug Output

https://gist.github.com/NathanDotTo/22c53af50e5eace1836fed7206eeedc1

### Panic Output

NA

### Expected Behavior

The dnac.ova file should have been uploaded to the s3 bucket.

### Actual Behavior

The dnac.ova file was not uploaded to the s3 bucket.

### Steps to Reproduce

`rm -rf .terraform*; terraform init; terraform apply`

### Important Factoids

See https://github.com/hashicorp/terraform/issues/28678#issuecomment-839167251

The dnac.ova file is ~80GB.

### References

https://github.com/hashicorp/terraform/issues/28678
https://github.com/hashicorp/terraform/issues/14693

**Comment by NathanDotTo** [comment]
Note that this code does work:

```
import boto3
from boto3.s3.transfer import TransferConfig

# Set the desired multipart threshold value (5GB)
GB = 1024 ** 3
config = TransferConfig(multipart_threshold=5*GB)

# Perform the transfer
s3 = boto3.client('s3')
s3.upload_file('/Users/x/Virtual Machines.localized/dnac.ova', 'dnac-ova', 'dnac/dnac.ova', Config=config)

```

**Comment by ewbankkit** [comment]
Related:

* #11656
* #788
* https://github.com/hashicorp/terraform/issues/23890

**Comment by ewbankkit** [comment]
@NathanDotTo Improvements to the Terraform file hashing functions (including `filemd5`) were made in https://github.com/hashicorp/terraform/pull/28681 and released in Terraform [**v0.15.4**](https://github.com/hashicorp/terraform/blob/v0.15/CHANGELOG.md#0154-may-19-2021).
Could you please upgrade to that version of the Terraform CLI, if possible, and retry your scenario?

**Comment by leonfs** [comment]
Are there any updates on this Issue? there seems to be a PR open for this fix.
